### ISSUE TITLE
tetGen meshing error 

### ISSUE THREAD

**Issue by BethOK** [gibbonCode/GIBBON#52]
Hi, I'm trying to run tetGen on a segment of a femur from an imported STL file, however I keep getting this error. 

"Found two nearly self-intersecting facets. 
  1st: [8938, 8981, 9339] #1 
  2nd: [8938, 8981, 9440] #1 
The dihedral angle between them is 0.0939087 degree. 
Hint:  You may use -p/# to decrease the dihedral angle  tolerance 0.1 (degree). 
A self-intersection was detected. Program stopped. 
Hint: use -d option to detect all self-intersections. 
--- Done --- 10-Sep-2019 14:51:12". 

How can I stop self-intersection? 

My inputs are: 
% define face boundary markers
    faceBoundaryMarker = ones(size(F,1),1);

% define region points
    V_regions = [0 0 0];
    V_holes = [];

% regional mesh volume parameter
    [regionA] = tetVolMeanEst(F,V); %Volume for regular tets

% create input structure 
    stringOpt = '-pq1.2AaY';

    inputStruct.stringOpt=stringOpt;
    inputStruct.Faces=F;
    inputStruct.Nodes=V;
    inputStruct.holePoints=V_holes;
    inputStruct.faceBoundaryMarker=faceBoundaryMarker; %Face boundary markers
    inputStruct.regionPoints=V_regions; %region points
    inputStruct.regionA=regionA;
    inputStruct.minRegionMarker=2; %Minimum region marker
    inputStruct.modelName=modelName;

% mesh model using tetrahedral elements using tetgen 
    [meshOutput]=runTetGen(inputStruct); %Run tetGen

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks for posting this

Can you check the following: 
- After importing the STL file all vertices are not shared, i.e. all triangles have their own triplet of points. To merge the points you need to do: 
```matlab
[F,V]=mergeVertices(F,V); %Point inside patch data defined by F and V
```
Did you do this? 
- After merging the points, if it is a closed shape, there should be no "boundary edges", you can check this with: 
```matlab
Eb=patchBoundary(F,V); %Boundary edges
```
If `Eb` is empty (`isempty(Eb)`) the surface is fully closed. Can you check if it is? You can plot the boundary points if there are any (find them using `indBoundary=unique(Eb(:))`) to try to understand your problem better. 
- The region point, `V_regions = [0 0 0];` needs to be inside the shape you wish to mesh. Is this the case? Can you plot it and check if `[0 0 0]` is indeed inside? If not you need to find a better interior point to serve as a region point. You can use something like this to get a suitable point (plot the point to verify it is inside):
```matlab
[V_regions]=getInnerPoint(F,V);
```
Here `F` represents your faces, and `V` your vertices. 
- If the problem persists, can you plot the coordinates listed in the error message to see the "violating" points? It currently says:
```
1st: [8938, 8981, 9339] 
2nd: [8938, 8981, 9440]
```
- If after the above the problem persists, you could share the file perhaps and I can have a look at it.

**Comment by BethOK** [comment]
Hi, 

Thanks for your response, unfortunately I cannot share the file with you due to privacy issues with the model. 
I tried all of your suggestions and my Eb showed no boundary edges (I had also already merged the vertices). 
However, as you pointed out, 0, 0, 0 did not lie within the shape. When I used the 'getInnerPoint' function, the function would get stuck on the  '%Creating points on faces' section, line 141 . I tried using a point that I know was inside the model, [60, -60, 60], however I still come up with the same meshing issue. 
Is there any reason why the 'getInnerPoint' function would not run? Does this imply there is a problem with my model? 
Then when I plotted the points where there is an intersection from error message: 
"Found two nearly self-intersecting facets. 
  1st: [8938, 8981, 9339] #1 
  2nd: [8938, 8981, 9440] #1 "
They do not lie within the model. 
Below I've copied out my meshing section of code, any help would be greatly appreciated. 
Thank you

% import a txt type STL file as patch data
    [stlStruct] = import_STL(fileName);
    
% access the data
    F = stlStruct.solidFaces{1}; %Faces
    V = stlStruct.solidVertices{1}; %Vertices

% In STL files nodes are not shared between triangles. Therefore the need to be merged with they are intended to be shared.
   [F,V] = mergeVertices(F,V); % Merging nodes
    
   Eb = patchBoundary(F,V); % Check that there are no boundary edges 
   if isempty(Eb) == 0
       fprintf('Boundary edge detected'); 
       return 
   end 

% define face boundary markers
    faceBoundaryMarker = ones(size(F,1),1);

% define reegion points
    V_regions = [60, -60, 60]; % getInnerPoint(F,V) not working;
    V_holes = [];

% regional mesh volume parameter
    [regionA] = tetVolMeanEst(F,V); %Volume for regular tets

% create input structure 
    stringOpt = '-pq1.2AaY';

    inputStruct.stringOpt=stringOpt;
    inputStruct.Faces=F;
    inputStruct.Nodes=V;
    inputStruct.holePoints=V_holes;
    inputStruct.faceBoundaryMarker=faceBoundaryMarker; %Face boundary markers
    inputStruct.regionPoints=V_regions; %region points
    inputStruct.regionA=regionA;
    inputStruct.minRegionMarker=2; %Minimum region marker
    inputStruct.modelName=modelName;

% mesh model using tetrahedral elements using tetgen 
    [meshOutput]=runTetGen(inputStruct); %Run tetGen
    
% access model element at patch data 
    Fb=meshOutput.facesBoundary;
    Cb=meshOutput.boundaryMarker;
    V=meshOutput.nodes;
    CE=meshOutput.elementMaterialID;
    E=meshOutput.elements;
    
% visualise 
    meshView(meshOutput,[]);
